e2e/api/api_test.go

```diff
@@ -0,0 +1,235 @@
+			},
+		}),
+		Entry("invalid IP", &types.VirtualService{
+			Id: "service1",
+			Key: &types.VirtualService_Key{
+				Ip:       "999.999.999.999",
+				Port:     8080,
+				Protocol: types.Protocol_TCP,
+			},
+			Config: &types.VirtualService_Config{
+				Scheduler: "sh",
+			},
+		}),
+		Entry("zero port", &types.VirtualService{
+			Id: "service1",
+			Key: &types.VirtualService_Key{
+				Ip:       "127.0.0.1",
+				Port:     0,
+				Protocol: types.Protocol_TCP,
+			},
+			Config: &types.VirtualService_Config{
+				Scheduler: "sh",
+			},
+		}),
+		Entry("missing protocol", &types.VirtualService{
+			Id: "service1",
+			Key: &types.VirtualService_Key{
+				Ip:   "127.0.0.1",
+				Port: 8080,
+			},
+			Config: &types.VirtualService_Config{
+				Scheduler: "sh",
+			},
+		}),
+		Entry("invalid protocol", &types.VirtualService{
+			Id: "service1",
+			Key: &types.VirtualService_Key{
+				Ip:       "127.0.0.1",
+				Port:     8080,
+				Protocol: 999,
+			},
+			Config: &types.VirtualService_Config{
+				Scheduler: "sh",
+			},
+		}),
+		Entry("config required", &types.VirtualService{
+			Id: "service1",
+			Key: &types.VirtualService_Key{
+				Ip:       "127.0.0.1",
+				Port:     8080,
+				Protocol: types.Protocol_TCP,
+			},
+		}),
+		Entry("scheduler required", &types.VirtualService{
+			Id: "service1",
+			Key: &types.VirtualService_Key{
+				Ip:       "127.0.0.1",
+				Port:     8080,
+				Protocol: types.Protocol_TCP,
+			},
+			Config: &types.VirtualService_Config{},
+		}),
+	)
+
+	DescribeTable("CreateServer", func(server *types.RealServer) {
+		_, err := client.CreateServer(ctx, server)
+		status, ok := status.FromError(err)
+
+		Expect(ok).To(BeTrue(), "got grpc status error")
+		if ok {
+			Expect(status.Code()).To(Equal(codes.InvalidArgument),
+				"expected InvalidArgument, but got %v", err)
+		}
+	},
+		Entry("empty server", &types.RealServer{}),
+		Entry("missing serviceID", &types.RealServer{
+			Key: &types.RealServer_Key{
+				Ip:   "172.16.1.1",
+				Port: 9090,
+			},
+			Config: &types.RealServer_Config{
+				Weight:  &wrappers.UInt32Value{Value: 2},
+				Forward: types.ForwardMethod_ROUTE,
+			},
+		}),
+		Entry("missing key", &types.RealServer{
+			ServiceID: "service1",
+			Config: &types.RealServer_Config{
+				Weight:  &wrappers.UInt32Value{Value: 2},
+				Forward: types.ForwardMethod_ROUTE,
+			},
+		}),
+		Entry("missing ip", &types.RealServer{
+			ServiceID: "service1",
+			Key: &types.RealServer_Key{
+				Port: 9090,
+			},
+			Config: &types.RealServer_Config{
+				Weight:  &wrappers.UInt32Value{Value: 2},
+				Forward: types.ForwardMethod_ROUTE,
+			},
+		}),
+		Entry("invalid ip", &types.RealServer{
+			ServiceID: "service1",
+			Key: &types.RealServer_Key{
+				Ip:   "999.999.999.999",
+				Port: 9090,
+			},
+			Config: &types.RealServer_Config{
+				Weight:  &wrappers.UInt32Value{Value: 2},
+				Forward: types.ForwardMethod_ROUTE,
+			},
+		}),
+		Entry("missing port", &types.RealServer{
```

Comment:
Should we also have a test for invalid port values? In `types.RealServer_Config`, it's defined as a `uint32`, but port numbers are inherently `uint16`.